CFE-3655: ansible promise type for running playbooks

tranchitella opened the pull request.

tranchitella commented:
@olehermanse ping

olehermanse commented on promise_types/ansible/README.md:
```diff
@@ -0,0 +1,79 @@
+| `playbook` | `string` | Absolute path of the Ansible playbook | No | Promiser |
+| `inventory` | `string` | Absolute path of the inventory file | No | - |
+| `limit` | `slist` | List of host names to target | No | `{"localhost"}` |
+| `tags` | `slist` | List of tags to play | no | `{}` |
```
```suggestion
| `tags` | `slist` | List of tags to play | No | `{}` |
```

olehermanse commented on promise_types/ansible/README.md:
```diff
@@ -0,0 +1,79 @@
+| `inventory` | `string` | Absolute path of the inventory file | No | - |
+| `limit` | `slist` | List of host names to target | No | `{"localhost"}` |
+| `tags` | `slist` | List of tags to play | no | `{}` |
+| `become` | `boolean` | Set the `become` option | - | `False` |
+| `become_method` | `string` | Set the `become_method` option | - | `"sudo"` |
+| `become_user` | `string` | Set the `become_user` option | - | `root` |
+| `connection` | `string` | Set the `connection` option; possible values: `local`, `ssh` | - | `local` |
+| `forks` | `int` | Set the `forks` option | - | `1` |
+| `private_key_file` | `string` | Absolute path of the SSH private key to use | - | - |
+| `remote_user` | `string` | Set the `remote_user` option | - | `root` |
```
Why do these have ` - ` for Mandatory column while the ones above have `No` ?

tranchitella replied:
It's a mistake, I'm fixing it in the other READMEs as well.

olehermanse commented on promise_types/ansible/README.md:
```diff
@@ -0,0 +1,79 @@
+
+This software was created by the team at [Northern.tech AS](https://northern.tech), with many contributions from the community. Thanks everyone!
+
+[CFengine](https://cfengine.com) is sponsored by [Northern.tech AS](https://northern.tech)
```
```suggestion
[CFEngine](https://cfengine.com) is sponsored by [Northern.tech AS](https://northern.tech)
```

olehermanse commented on promise_types/ansible/ansible_promise.py:
```diff
@@ -0,0 +1,163 @@
+    ANSIBLE_AVAILABLE = False
+
+
+COMPAT_OPTIONS = (
```
Maybe a comment to explain what COMPAT_OPTIONS means? Compatibility with what?

tranchitella replied:
sorry, it was a left-over

olehermanse commented on promise_types/ansible/example.cf:
```diff
@@ -0,0 +1,15 @@
+{
+  ansible:
+    "my_playbook"
+      playbook  => "/tmp/playbook.yaml",
+      inventory => "/tmp/inventory.yaml",
```
Maybe you could add a small dummy playbook to make it complete and the example runnable?

olehermanse commented on promise_types/ansible/ansible_promise.py:
```diff
@@ -0,0 +1,163 @@
+
+        def must_be_absolute(v):
+            if not os.path.isabs(v):
+                raise ValidationError(f"must be an absolute path, not '{v}'")
```
```suggestion
                raise ValidationError(f"Must be an absolute path, not '{v}'")
```

olehermanse reviewed (approved):
Looks really good, Fabio! I haven't tested it, but here are some smaller suggestions.

olehermanse commented:
@tranchitella your change to cfengine.py was already merged here: https://github.com/cfengine/modules/pull/10

You can rebase and drop that commit :)

tranchitella commented:
@olehermanse I addressed your comments, rebased and fixed the other README files.
Can you please review it again and merge, in case it is okay? I don't have merge rights in this repo.
Thanks.

vpodzime commented on promise_types/ansible/ansible_promise.py:
```diff
@@ -0,0 +1,164 @@
+    from ansible.module_utils.common.collections import ImmutableDict
+    from ansible.parsing.dataloader import DataLoader
+    from ansible.plugins.callback import CallbackBase
+    from ansible.vars.manager import VariableManager
```
Please only do the imports in the `try-except` block. The class definition shouldn't be inside it.

tranchitella replied:
The class depends on CallbackBase, which is imported conditionally. This is the reason why the class definition is in the try block. This is a common pattern in Python, I believe.

vpodzime replied:
The class can be defined inside an `if ANSIBLE_AVAILABLE`.

tranchitella replied:
ok done

vpodzime commented on promise_types/ansible/ansible_promise.py:
```diff
@@ -0,0 +1,164 @@
+
+    def evaluate_promise(
+        self, safe_promiser: str, attributes: Dict
+    ) -> Tuple[str, List[str]]:
```
This formatting breaks the flow of the code. I think this would be better:
```
def evaluate_promise(self, safe_promise: str, attributes: Dict)
                                   -> Tuple[str, List[str]]:
```
or even with everything on one line. Nobody's going to print this code on a toilet paper.

tranchitella replied:
The code is automatically formatted using black to avoid formatting bikeshedding.

tranchitella replied:
I'm use black to format my Python code, if the policy in CFEngine is not relying on autoformatter let me know and I'll manually format it according to your policy.

olehermanse replied:
@tranchitella black is better ;) Ignore him, your formatting is perfect.

olehermanse replied:
@vpodzime The problem is that you end up arguing something extremely subjective. I don't print my code on toilet paper and I still think Fabios snippet here is more readable than your suggestion.

vpodzime replied:
OK. If **you guys** prefer such formatting, that's the way to go. But not just because some formatter puts it that way. Formatting preferences differ, for me "everything after `def something` should be indented more than the `def`" is quite a key for readability and flow of the code. But YMMV, of course.

vpodzime commented on promise_types/ansible/ansible_promise.py:
```diff
@@ -0,0 +1,164 @@
+            passwords={},
+        )
+        callback = CallbackModule(promise=self)
+        pbex._tqm._stdout_callback = callback
```
Is there a way to avoid accessing private attributes?

tranchitella replied:
> Please split the changes in the _git_ module into a separate commit.

done

tranchitella replied:
> Is there a way to avoid accessing private attributes?

Yes, but it would require bloating the Python code with all the initialization of the task queue manager which the PlaybookExecutor embeds. In any case, _stdout_callback is private, thus we would need to override the TaskQueueManager, adding additional complexity. I don't think it's a good idea, but if you disagree I can give it a try.

vpodzime replied:
Not worth the struggle, AFAICT. We just need to hope the names stay the same.

vpodzime reviewed (changes requested):
Please split the changes in the *git* module into a separate commit.
